modules/renter/contractor/journal.go

```diff
@@ -0,0 +1,332 @@
+	Data     rawJSON     `json:"d"`
+	Checksum crypto.Hash `json:"c"`
+}
+
+// TODO: replace with json.RawMessage after upgrading to Go 1.8
+type rawJSON []byte
+
+// MarshalJSON returns r as the JSON encoding of r.
+func (r rawJSON) MarshalJSON() ([]byte, error) {
+	if r == nil {
+		return []byte("null"), nil
+	}
+	return r, nil
+}
+
+// UnmarshalJSON sets *r to a copy of data.
+func (r *rawJSON) UnmarshalJSON(data []byte) error {
+	if r == nil {
+		return errors.New("rawJSON: UnmarshalJSON on nil pointer")
+	}
+	*r = append((*r)[:0], data...)
+	return nil
+}
+
+type updateSet []journalUpdate
+
+func (set updateSet) MarshalJSON() ([]byte, error) {
+	marshaledSet := make([]marshaledUpdate, len(set))
+	for i, u := range set {
+		data, err := json.Marshal(u)
+		if err != nil {
+			build.Critical("failed to marshal known type:", err)
+		}
+		marshaledSet[i].Data = data
+		marshaledSet[i].Checksum = crypto.HashBytes(data)
+		switch u.(type) {
+		case updateUploadRevision:
+			marshaledSet[i].Type = "uploadRevision"
+		case updateDownloadRevision:
+			marshaledSet[i].Type = "downloadRevision"
+		case updateCachedUploadRevision:
+			marshaledSet[i].Type = "cachedUploadRevision"
+		case updateCachedDownloadRevision:
+			marshaledSet[i].Type = "cachedDownloadRevision"
+		}
+	}
+	return json.Marshal(marshaledSet)
+}
+
+func (set *updateSet) UnmarshalJSON(b []byte) error {
+	var marshaledSet []marshaledUpdate
+	if err := json.Unmarshal(b, &marshaledSet); err != nil {
+		return err
+	}
+	for _, u := range marshaledSet {
+		if crypto.HashBytes(u.Data) != u.Checksum {
+			return errors.New("bad checksum")
+		}
+		var err error
+		switch u.Type {
+		case "uploadRevision":
+			var ur updateUploadRevision
+			err = json.Unmarshal(u.Data, &ur)
+			*set = append(*set, ur)
+		case "downloadRevision":
+			var dr updateDownloadRevision
+			err = json.Unmarshal(u.Data, &dr)
+			*set = append(*set, dr)
+		case "cachedUploadRevision":
+			var cur updateCachedUploadRevision
+			err = json.Unmarshal(u.Data, &cur)
+			*set = append(*set, cur)
+		case "cachedDownloadRevision":
+			var cdr updateCachedDownloadRevision
+			err = json.Unmarshal(u.Data, &cdr)
+			*set = append(*set, cdr)
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type updateUploadRevision struct {
+	NewRevisionTxn     types.Transaction `json:"newrevisiontxn"`
+	NewSectorRoot      crypto.Hash       `json:"newsectorroot"`
+	NewSectorIndex     int               `json:"newsectorindex"`
+	NewUploadSpending  types.Currency    `json:"newuploadspending"`
+	NewStorageSpending types.Currency    `json:"newstoragespending"`
+}
+
+func (u updateUploadRevision) apply(data *contractorPersist) {
+	if len(u.NewRevisionTxn.FileContractRevisions) == 0 {
+		build.Critical("updateUploadRevision is missing its FileContractRevision")
+		return
+	}
+	rev := u.NewRevisionTxn.FileContractRevisions[0]
+	c := data.Contracts[rev.ParentID.String()]
+	if u.NewSectorIndex != len(c.MerkleRoots) {
```

Comment:
`if u.NewSectorIndes > len(c.MerkleRoots) {`

if we break it into multiple files, it's possible that the merkle roots are longer at this point, but we don't catch up until the future updates. Best to fix the sanity check to reflect that situation now, even though it technically can't happen in the single-file scenario